File: app/src/main/java/com/sstechcanada/todo/activities/TodoListActivity.java
package com.sstechcanada.todo.activities;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.databinding.DataBindingUtil;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.sstechcanada.todo.R;
import com.sstechcanada.todo.activities.auth.LoginActivity;
import com.sstechcanada.todo.adapters.TodoListAdapter;
import com.sstechcanada.todo.broadcast_receivers.DailyAlarmReceiver;
import com.sstechcanada.todo.data.TodoListContract;
import com.sstechcanada.todo.data.TodoListDbHelper;
import com.sstechcanada.todo.data.TodoListProvider;
import com.sstechcanada.todo.databinding.ActivityTodoListBinding;
import com.sstechcanada.todo.models.TodoTask;
import com.sstechcanada.todo.utils.NotificationUtils;
import com.sstechcanada.todo.utils.SaveSharedPreference;

public class TodoListActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<Cursor>,
        TodoListAdapter.TodoListAdapterOnClickHandler,
        SharedPreferences.OnSharedPreferenceChangeListener {
    private static final String TAG = TodoListActivity.class.getSimpleName();
    private static final int ADD_TASK_REQUEST = 1;
    private static final int EDIT_TASK_REQUEST = 2;
    private static final int ID_TODOLIST_LOADER = 2018;
    String userID;
    private int list_limit = 1, db_cnt = 0;
    private RecyclerView mRecyclerView;
    private TodoListAdapter mTodoListAdapter;
    private ActivityTodoListBinding mBinding;
    private SharedPreferences mSharedPreferences, ll;
    private AppCompatImageView toolbar_profile;
    private TodoListDbHelper tld;
    private FirebaseAuth mAuth;
    private FirebaseUser user;
    private DatabaseReference databaseReference;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_todo_list);
        mRecyclerView = mBinding.rvTodoList;
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        layoutManager.setReverseLayout(true);
        layoutManager.setStackFromEnd(true);
        mRecyclerView.setLayoutManager(layoutManager);
        mTodoListAdapter = new TodoListAdapter(this, this);
        mRecyclerView.setAdapter(mTodoListAdapter);

        mAuth = FirebaseAuth.getInstance();
        user = mAuth.getCurrentUser();
        tld = new TodoListDbHelper(TodoListActivity.this);

        //Limit Set
        db_cnt = tld.todoCount();
        setValue();

        showHidePlaceholder();


//        getSupportActionBar().setDisplayOptions(ActionBar.DISPLAY_SHOW_CUSTOM);
//        getSupportActionBar().setCustomView(R.layout.abs_layout);
//        DividerItemDecoration mDividerItemDecoration = new DividerItemDecoration(this, LinearLayoutManager.VERTICAL);
//        mRecyclerView.addItemDecoration(mDividerItemDecoration);

        toolbar_profile = findViewById(R.id.profile_toolbar);
        toolbar_profile.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(TodoListActivity.this, LoginActivity.class));
            }
        });

        FloatingActionButton fab = mBinding.fab;
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                db_cnt = tld.todoCount();
                setValue();
                if (isLogin()) {
                    Intent intent = new Intent(TodoListActivity.this, AddOrEditTaskActivity.class);
                    intent.putExtra(getString(R.string.intent_adding_or_editing_key), getString(R.string.add_new_task));
                    startActivityForResult(intent, ADD_TASK_REQUEST);
                }
            }
        });

        mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);

        getSupportLoaderManager().initLoader(ID_TODOLIST_LOADER, null, this);

        //scheduleDailyDueCheckerAlarm();
        //cancelAlarm();

        mBinding.completedTab.setOnClickListener(view -> {
            startActivity(new Intent(TodoListActivity.this, AppUpgradeActivity.class));
        });
    }

    private void showHidePlaceholder() {
        if (db_cnt == 0) {
            mBinding.placeholderImage.setVisibility(View.VISIBLE);
        } else {
            mBinding.placeholderImage.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.todo_list_menu, menu);
        //For 3 Dot menu
        return false;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                break;
            case R.id.action_delete_test:
                // later a service will take care of this
                deleteAllCheckedTasks();
                break;
            case R.id.action_test_something:
                // just a place for me to test whatever I'm testing at the moment
                NotificationUtils.notifyUserOfDueAndOverdueTasks(this);
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onClick(TodoTask todoTask, View view) {
        // are they checking or unchecking the task checkbox, or tapping the task to edit it?
        if (view instanceof CheckBox) {
            // checking off task gets it flagged for deletion soon, and unchecking it reprieves it
            final String id = String.valueOf(todoTask.getId());
            final Uri uri = TodoListContract.TodoListEntry.CONTENT_URI.buildUpon().appendPath(id).build();
            int isCompleted;

            final ContentValues contentValues = new ContentValues();
            contentValues.put(TodoListContract.TodoListEntry.COLUMN_DESCRIPTION, todoTask.getDescription());
            contentValues.put(TodoListContract.TodoListEntry.COLUMN_PRIORITY, todoTask.getPriority());
            contentValues.put(TodoListContract.TodoListEntry.COLUMN_DUE_DATE, todoTask.getDueDate());

            if (((CheckBox) view).isChecked()) {
                isCompleted = TodoTask.TASK_COMPLETED;
            } else {
                isCompleted = TodoTask.TASK_NOT_COMPLETED;
            }

            contentValues.put(TodoListContract.TodoListEntry.COLUMN_COMPLETED, isCompleted);

            // Wait half a second so they can briefly see the check appear or disappear before
            // the task is moved to or from the bottom
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                public void run() {
                    getContentResolver().update(uri, contentValues, "_id=?", new String[]{id});
                    updateWidget();
                }
            }, 500);
        } else {
            // edit the task
            Intent intent = new Intent(this, AddOrEditTaskActivity.class);
            intent.putExtra(getString(R.string.intent_adding_or_editing_key), getString(R.string.edit_task));
            intent.putExtra(getString(R.string.intent_todo_key), todoTask);
            startActivityForResult(intent, EDIT_TASK_REQUEST);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            updateWidget();
        }
    }

    private void updateWidget() {
        // let the widget know there's been a database or sort order change
        Intent intent = new Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        sendBroadcast(intent);
    }

    @Override
    public Loader<Cursor> onCreateLoader(int loaderId, Bundle bundle) {
        if (loaderId == ID_TODOLIST_LOADER) {
            String sortOrderPreference = getSortOrderPreference();
            String sortOrder;
            Uri todoListQueryUri = TodoListContract.TodoListEntry.CONTENT_URI;
            // sort order preference is the primary sort, with the other sort order as secondary
            if (sortOrderPreference.equals(getString(R.string.priority))) {
                sortOrder = TodoListProvider.SORT_ORDER_PRIORITY;
            } else {
                sortOrder = TodoListProvider.SORT_ORDER_DUEDATE;
            }

            return new CursorLoader(this,
                    todoListQueryUri,
                    null,
                    null,
                    null,
                    sortOrder);
        } else {
            throw new RuntimeException("Loader Not Implemented: " + loaderId);
        }
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        mTodoListAdapter.swapCursor(data);
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> loader) {
    }

    private String getSortOrderPreference() {
        return mSharedPreferences.getString(getString(R.string.pref_sort_by_key), getString(R.string.priority));
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
        mTodoListAdapter.swapCursor(null);
        getSupportLoaderManager().restartLoader(ID_TODOLIST_LOADER, null, this);
        updateWidget();
    }

    @Override
    protected void onResume() {

        super.onResume();
        // This is so that if we've edited a task directly from the widget, the widget will still
        // get updated when we come to this activity after clicking UPDATE TASK in AddOrEditTaskActivity
        updateWidget();
        showHidePlaceholder();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mSharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
    }

    private void deleteAllCheckedTasks() {
        // this is just for testing, later a service will do it periodically
        Uri uri = TodoListContract.TodoListEntry.CONTENT_URI;
        int deletedTasksCount = getContentResolver().delete(uri, "completed=?", new String[]{String.valueOf(TodoTask.TASK_COMPLETED)});
        if (deletedTasksCount > 0) {
            updateWidget();
        }
    }

    public void scheduleDailyDueCheckerAlarm() {
        Intent intent = new Intent(getApplicationContext(), DailyAlarmReceiver.class);

        final PendingIntent pendingIntent = PendingIntent.getBroadcast(this,
                DailyAlarmReceiver.REQUEST_CODE,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT);
        long firstMillis = System.currentTimeMillis(); // alarm is set right away

        AlarmManager alarm = (AlarmManager) this.getSystemService(Context.ALARM_SERVICE);
        // First parameter is the type: ELAPSED_REALTIME, ELAPSED_REALTIME_WAKEUP, RTC_WAKEUP
        // Interval can be INTERVAL_FIFTEEN_MINUTES, INTERVAL_HALF_HOUR, INTERVAL_HOUR, INTERVAL_DAY
        alarm.setInexactRepeating(AlarmManager.RTC_WAKEUP, firstMillis,
                AlarmManager.INTERVAL_FIFTEEN_MINUTES, pendingIntent);
    }

    public void cancelAlarm() {
        Intent intent = new Intent(getApplicationContext(), DailyAlarmReceiver.class);
        final PendingIntent pIntent = PendingIntent.getBroadcast(this, DailyAlarmReceiver.REQUEST_CODE,
                intent, PendingIntent.FLAG_UPDATE_CURRENT);
        AlarmManager alarm = (AlarmManager) this.getSystemService(Context.ALARM_SERVICE);
        alarm.cancel(pIntent);
    }

    public boolean isLogin() {
        FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
        if (user == null) {
            Toast.makeText(this, "You need to login ", Toast.LENGTH_SHORT).show();
            return false;
        } else if (list_limit <= db_cnt) {
            //Limit Check
            Toast.makeText(this, "You can not store more than " + list_limit, Toast.LENGTH_SHORT).show();
            return false;
        }
        return true;
    }

    public void setValue() {
        if (user != null) {
            list_limit = SaveSharedPreference.loadLimit(this);
        }
//        Toast.makeText(this, list_limit + " " + db_cnt, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/sstechcanada/todo/activities/TodoListActivity.kt
package com.sstechcanada.todo.activities

import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Canvas
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import android.view.WindowManager
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.res.ResourcesCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import com.airbnb.lottie.LottieAnimationView
import com.bumptech.glide.Glide
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.getkeepsafe.taptargetview.TapTarget
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.sstechcanada.todo.R
import com.sstechcanada.todo.activities.auth.LoginActivity
import com.sstechcanada.todo.activities.auth.ProfileActivity
import com.sstechcanada.todo.adapters.TodoListFirestoreAdapter
import com.sstechcanada.todo.databinding.ActivityTodoListBinding
import com.sstechcanada.todo.models.TodoTaskFirestore
import com.sstechcanada.todo.utils.SaveSharedPreference
import com.sstechcanada.todo.utils.SwipeController
import com.sstechcanada.todo.utils.SwipeControllerActions
import es.dmoral.toasty.Toasty
import kotlinx.android.synthetic.main.act_bar.*
import kotlinx.android.synthetic.main.activity_todo_list.*

class TodoListActivity : AppCompatActivity() {
    private val db = FirebaseFirestore.getInstance()
    private val usersColRef = db.collection("Users")
    var userID: String? = null
    var loadingProgressBar: ProgressBar? = null
    var fab: FloatingActionButton? = null
    private var list_limit = 15
    private var mBinding: ActivityTodoListBinding? = null
    private var toolbar_profile: AppCompatImageView? = null
    private var mAuth: FirebaseAuth? = null
    private var user: FirebaseUser? = null
    private var todoListFirestoreAdapter: TodoListFirestoreAdapter? = null
    var editor: SharedPreferences.Editor? = null
    private var mInterstitialAd: InterstitialAd? = null
    private var listId: String? = null
    private var listName: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_todo_list)
        loadingProgressBar = mBinding?.loadingProgressBar
        fab = mBinding?.fab
        if (Integer.valueOf(MasterTodoListActivity.purchaseCode) == 1 || Integer.valueOf(MasterTodoListActivity.purchaseCode) == 2) {
            Log.i("purchase code", MasterTodoListActivity.purchaseCode)
            Log.i("purchase code", "purchaseCode")
            mBinding?.completedTab?.setCompoundDrawablesWithIntrinsicBounds(
                0,
                0,
                R.drawable.ic_baseline_check_circle_24,
                0
            )
        }
        val prefs = getSharedPreferences(LoginActivity.SHAREDPREF, MODE_PRIVATE)
        editor = getSharedPreferences(LoginActivity.SHAREDPREF, MODE_PRIVATE).edit()
        showProgressBar()
        lottieAnimationView = findViewById(R.id.placeholderImage)
        mAuth = FirebaseAuth.getInstance()
        user = mAuth!!.currentUser
        if (user != null) {
            userID = user!!.uid
        }
        if (prefs.getBoolean("flagTodoListFirstRun", true)) {
            mBinding?.buttonTapTargetView?.visibility = View.INVISIBLE
            callWalkThrough()
        }
        setupObservers()
        val adView = mBinding?.adView
        if (SaveSharedPreference.getAdsEnabled(this)) {
            loadFullScreenAds()
            val adRequest = AdRequest.Builder().build()
            adView?.loadAd(adRequest)
        } else {
            adView?.visibility = View.GONE
        }
        toolbar_profile = findViewById(R.id.profile_toolbar)
        Glide.with(this).load(mAuth?.currentUser?.photoUrl).into(profile_toolbar)
        toolbar_profile?.setOnClickListener {
            startActivity(
                Intent(
                    this@TodoListActivity,
                    ProfileActivity::class.java
                )
            )
        }
        fab?.setOnClickListener {
            db_cnt = todoListFirestoreAdapter?.itemCount ?: 0
            Log.i("ItemCount", "FAB Clicked")
            if (LoginActivity.userAccountDetails[1].toInt() > (todoListFirestoreAdapter?.itemCount ?: 0)) {
                setValue()
                if (isLogin) {
                    val intent = Intent(this@TodoListActivity, AddOrEditTaskActivity::class.java)
                    intent.putExtra(
                        getString(R.string.intent_adding_or_editing_key),
                        getString(R.string.add_new_task)
                    )
                    startActivityForResult(intent, ADD_TASK_REQUEST)
                }
            } else {
                if (isLogin) {
                    if (MasterTodoListActivity.purchaseCode != "2") {
                        val intent = Intent(this@TodoListActivity, AppUpgradeActivity::class.java)
                        //                        intent.putExtra(getString(R.string.intent_adding_or_editing_key), getString(R.string.add_new_task));
                        Toasty.info(
                            applicationContext,
                            getString(R.string.upgrade_master_list),
                            Toast.LENGTH_SHORT
                        ).show()
                        startActivity(intent)
                    } else if (MasterTodoListActivity.purchaseCode == "2") {
                        Toasty.warning(
                            applicationContext,
                            "Sorry, You cannot add more to-do items. You have reached the max-limit!",
                            Toast.LENGTH_LONG
                        ).show()
                    }
                }
            }
        }
        mBinding?.completedTab?.setOnClickListener {
            if (Integer.valueOf(MasterTodoListActivity.purchaseCode) == 1 || Integer.valueOf(MasterTodoListActivity.purchaseCode) == 2) {
                startActivity(Intent(this@TodoListActivity, CompletedTodoListActivity::class.java))
            } else {
                startActivity(Intent(this@TodoListActivity, AppUpgradeActivity::class.java))
            }
        }
        rv_todo_list.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 12 || dy < 0 && fab!!.isShown) {
                    fab?.hide()
                }
                if (dy < -12 && !fab!!.isShown) {
                    fab?.show()
                }
                if (!recyclerView.canScrollVertically(-1)) {
                    fab?.show()
                }
            }
        })
    }

    private fun setupObservers() {
        fetchIntent()
        //Limit Set
        setValue()
    }

    private fun fetchIntent() {
        if (intent.extras != null) {
            listName = intent.getStringExtra("ListName")
            listId = intent.getStringExtra("ListId")
            if (listId != null && listName != null) {
                setUpFirestoreRecyclerView()
            }
        }
    }

    private fun setUpFirestoreRecyclerView() {
//        listId=getIntent().getStringExtra("ListId");
        Log.i("ListId", "Setupdrecyasdf")
        Log.i("ListId", "n$listId")
        FirebaseCrashlytics.getInstance().log(this.javaClass.simpleName + "listId = " + listId)
        FirebaseCrashlytics.getInstance().log(this.javaClass.simpleName + "UserId = " + userID)
        val query = listId?.let {
            userID?.let { it1 ->
                usersColRef.document(it1).collection("Lists").document(
                    it
                ).collection("Todo").whereEqualTo("Status", "Pending")
                    .orderBy("priority", Query.Direction.DESCENDING)
            }
        }
        val options = query?.let {
            FirestoreRecyclerOptions.Builder<TodoTaskFirestore>()
                .setQuery(it, TodoTaskFirestore::class.java).build()
        }
        todoListFirestoreAdapter = options?.let { TodoListFirestoreAdapter(it, this) }
        rv_todo_list.setHasFixedSize(true)
        rv_todo_list.layoutManager = LinearLayoutManager(this)
        val swipeController = SwipeController(this, object : SwipeControllerActions() {
            override fun onRightClicked(position: Int) {
                Log.i("click", "right")
                if (SaveSharedPreference.getAdsEnabled(this@TodoListActivity)) {
                    loadFullScreenAds()
                }
                AlertDialog.Builder(this@TodoListActivity)
                    .setIcon(android.R.drawable.ic_menu_delete)
                    .setTitle("Confirm Delete")
                    .setMessage("Are you sure you want to delete this task?")
                    .setPositiveButton("Yes") { dialog: DialogInterface?, which: Int ->
                        if (SaveSharedPreference.getAdsEnabled(this@TodoListActivity)) {
                            if (mInterstitialAd != null) {
                                mInterstitialAd?.show(this@TodoListActivity)
                                val documentSnapshot =
                                    todoListFirestoreAdapter?.snapshots?.getSnapshot(position)
                                val id = documentSnapshot?.id
                                userID?.let {
                                    listId?.let { it1 ->
                                        id?.let { it2 ->
                                            usersColRef.document(it).collection("Lists")
                                                .document(it1).collection("Todo").document(it2)
                                                .delete()
                                        }
                                    }
                                }
                                Toasty.error(
                                    this@TodoListActivity,
                                    "Task Deleted",
                                    Toast.LENGTH_SHORT
                                ).show()
                            } else {
                                val documentSnapshot =
                                    todoListFirestoreAdapter!!.snapshots.getSnapshot(position)
                                val id = documentSnapshot.id
                                usersColRef.document(userID!!).collection("Lists")
                                    .document(listId!!).collection("Todo").document(id).delete()
                                Toasty.error(
                                    this@TodoListActivity,
                                    "Task Deleted",
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        } else {
                            val documentSnapshot =
                                todoListFirestoreAdapter?.snapshots?.getSnapshot(position)
                            val id = documentSnapshot?.id
                            if (id != null) {
                                userID?.let {
                                    usersColRef.document(it).collection("Lists").document(listId!!)
                                        .collection("Todo").document(id).delete()
                                }
                            }
                            Toasty.error(this@TodoListActivity, "Task Deleted", Toast.LENGTH_SHORT)
                                .show()
                        }
                    }
                    .setNegativeButton("No", null)
                    .show()
            }

            override fun onLeftClicked(position: Int) {
                val documentSnapshot = todoListFirestoreAdapter!!.snapshots.getSnapshot(position)
                val id = documentSnapshot.id
                val task = documentSnapshot.toObject(
                    TodoTaskFirestore::class.java
                )
                var todoTask: TodoTaskFirestore? = null
                if (task != null) {
                    todoTask = TodoTaskFirestore(
                        task.description,
                        task.priority,
                        task.dueDate,
                        id,
                        task.status,
                        task.category,
                        2,
                        task.benefits,
                        task.benefitsString,
                        task.timestampCompleted
                    )
                }
                val intent = Intent(this@TodoListActivity, AddOrEditTaskActivity::class.java)
                intent.putExtra("Adding or editing", "Edit Task")
                intent.putExtra("Todo", todoTask)
                startActivity(intent)
            }
        })
        val itemTouchhelper = ItemTouchHelper(swipeController)
        itemTouchhelper.attachToRecyclerView(rv_todo_list)
        rv_todo_list.adapter = todoListFirestoreAdapter
        hideProgressBar()
        rv_todo_list.addItemDecoration(object : ItemDecoration() {
            override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
                swipeController.onDraw(c)
            }
        })
        db_cnt = todoListFirestoreAdapter!!.itemCount
    }

    override fun onStart() {
        super.onStart()
        if (todoListFirestoreAdapter != null) todoListFirestoreAdapter!!.startListening()
        mBinding!!.listNameTextView.text = listName
    }

    override fun onStop() {
        super.onStop()
        if (todoListFirestoreAdapter != null) todoListFirestoreAdapter!!.stopListening()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.todo_list_menu, menu)
        //For 3 Dot menu
        return false
    }

    //        else if (list_limit <= db_cnt) {
//            //Limit Check
//            Toasty.info(this, getString(R.string.upgrade_todo_list), Toast.LENGTH_LONG, true).show();
//            startActivity(new Intent(TodoListActivity2.this, AppUpgradeActivity2.class));
//            return false;
//        }
    private val isLogin: Boolean
        get() {
            val user = FirebaseAuth.getInstance().currentUser
            if (user == null) {
                Toasty.warning(this, getString(R.string.login_first), Toast.LENGTH_LONG).show()
                startActivity(Intent(this@TodoListActivity, LoginActivity::class.java))
                return false
            }
            //        else if (list_limit <= db_cnt) {
//            //Limit Check
//            Toasty.info(this, getString(R.string.upgrade_todo_list), Toast.LENGTH_LONG, true).show();
//            startActivity(new Intent(TodoListActivity2.this, AppUpgradeActivity2.class));
//            return false;
//        }
            return true
        }

    fun setValue() {
        if (user != null) {
            list_limit = 15
        }
    }

    private fun showProgressBar() {
        loadingProgressBar!!.visibility = View.VISIBLE
        window.setFlags(
            WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
            WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
        )
    }

    private fun hideProgressBar() {
        if (loadingProgressBar!!.visibility == View.VISIBLE) {
            loadingProgressBar!!.visibility = View.INVISIBLE
            window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        startActivity(Intent(this@TodoListActivity, MasterTodoListActivity::class.java))
    }

    private fun loadFullScreenAds() {
        val adRequest = AdRequest.Builder().build()
        //ca-app-pub-3111421321050812/5967628112 our
        //test ca-app-pub-3940256099942544/1033173712
        InterstitialAd.load(this, "ca-app-pub-3111421321050812/5967628112", adRequest,
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(interstitialAd: InterstitialAd) {
                    // The mInterstitialAd reference will be null until
                    // an ad is loaded.
                    mInterstitialAd = interstitialAd
                    mInterstitialAd!!.fullScreenContentCallback =
                        object : FullScreenContentCallback() {
                            override fun onAdDismissedFullScreenContent() {
                                // Called when fullscreen content is dismissed.
                                Log.d("TAG", "The ad was dismissed.")
                            }

                            override fun onAdFailedToShowFullScreenContent(adError: AdError) {
                                // Called when fullscreen content failed to show.
                                Log.d("TAG", "The ad failed to show.")
                            }

                            override fun onAdShowedFullScreenContent() {
                                // Called when fullscreen content is shown.
                                // Make sure to set your reference to null so you don't
                                // show it a second time.
                                mInterstitialAd = null
                                Log.d("TAG", "The ad was shown.")
                            }
                        }
                }

                override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                    // Handle the error
                    mInterstitialAd = null
                }
            })
    }

    private fun callWalkThrough() {
        TapTargetSequence(this)
            .targets(
                TapTarget.forView(fab, "Add Button", "Click here to add a new list item")
                    .outerCircleColor(R.color.chip_5)
                    .outerCircleAlpha(0.98f)
                    .targetCircleColor(R.color.colorUncompletedBackground)
                    .titleTextSize(22)
                    .titleTextColor(R.color.colorUncompletedBackground)
                    .descriptionTextSize(16)
                    .titleTypeface(ResourcesCompat.getFont(this, R.font.poppins_semibold))
                    .textTypeface(ResourcesCompat.getFont(this, R.font.raleway_medium))
                    .descriptionTextColor(R.color.black)
                    .textColor(R.color.black)
                    .dimColor(R.color.black)
                    .drawShadow(true)
                    .cancelable(false)
                    .tintTarget(true)
                    .transparentTarget(true)
                    .targetRadius(80),
                TapTarget.forView(
                    mBinding?.buttonTapTargetView,
                    "List Items",
                    "Swipe right and click on the pencil icon to edit a list item. " +
                            "\n\nSwipe left and click on the garbage can icon to delete a list item. " +
                            "\n\nThis app works better for one-time tasks (e.g. \"buy glasses\") than recurring ones (e.g. \"go to gym every day\")."
                )
                    .outerCircleColor(R.color.chip_5)
                    .outerCircleAlpha(0.98f)
                    .targetCircleColor(R.color.colorUncompletedBackground)
                    .titleTextSize(22)
                    .titleTextColor(R.color.colorUncompletedBackground)
                    .descriptionTextSize(16)
                    .titleTypeface(ResourcesCompat.getFont(this, R.font.poppins_semibold))
                    .textTypeface(ResourcesCompat.getFont(this, R.font.raleway_medium))
                    .descriptionTextColor(R.color.black)
                    .textColor(R.color.black)
                    .dimColor(R.color.black)
                    .drawShadow(true)
                    .cancelable(false)
                    .tintTarget(true)
                    .transparentTarget(true)
                    .targetRadius(60)
            ).listener(object : TapTargetSequence.Listener {
                override fun onSequenceFinish() {

//                Toast.makeText(TodoListActivity2.this,"Sequence Finished",Toast.LENGTH_SHORT).show();
                    Toasty.success(
                        this@TodoListActivity,
                        "You are all set now!",
                        Toast.LENGTH_SHORT
                    ).show()
                    //                flagTodoListFirstRun = false;
                    mBinding?.buttonTapTargetView?.visibility = View.GONE
                    editor?.putBoolean("flagTodoListFirstRun", false)
                    editor?.apply()
                }

                override fun onSequenceStep(lastTarget: TapTarget, targetClicked: Boolean) {}
                override fun onSequenceCanceled(lastTarget: TapTarget) {
                    editor?.putBoolean("flagTodoListFirstRun", false)
                    mBinding?.buttonTapTargetView?.visibility = View.GONE
                    editor?.apply()
                }
            }).start()
    }

    companion object {
        private const val ADD_TASK_REQUEST = 1
        var db_cnt = 0

        @JvmField
        var lottieAnimationView: LottieAnimationView? = null

        @JvmStatic
        fun showPlaceHolder() {
            lottieAnimationView?.visibility = View.VISIBLE
        }

        @JvmStatic
        fun hidePlaceHolder() {
            lottieAnimationView?.visibility = View.INVISIBLE
        }
    }
}